EducationalAssignments/SecureHashPartOne.md

```diff
@@ -0,0 +1,308 @@
+    python repy.py restrictions.default encasementlib.r2py [security_layer].r2py [attack_program].r2py 
+    
+    """ 
+TYPE="type"
+ARGS="args"
+RETURN="return"
+EXCP="exceptions"
+TARGET="target"
+FUNC="func"
+OBJC="objc"
+
+class SecureFile():
+	def __init__(self,filename,create):
+		# globals
+		mycontext['debug'] = False   
+		
+		# local (per object) reference to the underlying file
+		self.fn = filename
+		self.count_reads = 0
+		
+		self.file = openfile(self.fn,create)
+		self.readcountlock = createlock()
+
+	def writeat(self,data,offset):	
+		# Write the requested data to the file using the sandbox's writeat call
+		self.file.writeat(data, offset)		
+  
+	def readat(self,bytes,offset):
+		# Read from the file using the sandbox's readat...
+		retval = self.file.readat(bytes,offset)
+		self.readcountlock.acquire(True)
+		self.count_reads = self.count_reads + 1
+		self.readcountlock.release()
+		return retval
+
+	def close(self):
+		self.file.close()
+	
+	def count_readat(self): 
+		# do not modify this function 
+		# returns the number of times readat() has been called on a file object
+		return self.count_reads
+
+def secureopenfile(filename, create):
+	return SecureFile(filename,create)
+
+# The code here sets up type checking and variable hiding for you.  You
+# should not need to change anything below here.
+sec_file_def = {"obj-type":SecureFile,
+                "name":"SecureFile",
+                "writeat":{"type":"func","args":(str,(int,long)),"exceptions":Exception,"return":(int,type(None)),"target":SecureFile.writeat},
+                "readat":{"type":"func","args":((int,long,type(None)),(int,long)),"exceptions":Exception,"return":str,"target":SecureFile.readat},
+		"count_readat":{"type":"func","args":None,"exceptions":None,"return":(int,type(None)),"target":SecureFile.count_readat},
+                "close":{"type":"func","args":None,"exceptions":None,"return":(bool,type(None)),"target":SecureFile.close}
+           	}
+
+CHILD_CONTEXT_DEF["secureopenfile"] = {TYPE:OBJC,ARGS:(str,bool),EXCP:Exception,RETURN:sec_file_def,TARGET:secureopenfile}
+
+# Execute the user code
+secure_dispatch_module()
+```
+
+
+
+### Testing your security layer
+----
+In this part of the assignment you will pretend to be an attacker. Remember
+the attacker's objective is to bypass the parity restrictions or cause
+the security layer to act in a disallowed manner. By understanding how the
+attacker thinks, you will be able to write better security layers.  
+
+An example of a test / attack is found below:
+
+```
+if "trackme.foo" in listfiles():
+  removefile("trackme.foo")
+
+#Create a file
+myfile = secureopenfile("trackme.foo",True)  
+
+# Put some valid data in the file.
+myfile.writeat("AbCd",0)
+
+# I should be able to read it out.
+assert('AbCd' == myfile.readat(None, 0))
+
+# may read block 1 since it writes to block 1 and "hello" is
+# contained entirely in that block
+myfile.writeat("hello",123)
```

Comment:
This should fail, right?  The file is smaller than this

Reply:
Yes, I did some changes to this part.